merkletree/tree.go

```diff
@@ -0,0 +1,219 @@
+package merkletree
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"github.com/filecoin-project/go-data-segment/util"
+	"log"
+	"reflect"
+)
+
+const digestBytes = 32
+
+// MerkleTree represents a Merkle tree which can be used to construct proof of containment for either leafs, subtrees or a sequence of leafs (subtrees)
+type MerkleTree interface {
+	// Depth returns the Depth of the tree. A single-node tree has Depth 1
+	Depth() int
+	// LeafCount returns the amount of leafs in the Merkle tree
+	LeafCount() int
+	// Root returns the root node of the tree
+	Root() *Node
+	// Leafs returns all the leaf nodes in the tree
+	Leafs() []Node
+	// ConstructProof constructs a Merkle proof of the subtree (or leaf) at level lvl with index idx.
+	// level 0 is the root and index 0 is the left-most node in a level.
+	ConstructProof(lvl int, idx int) (MerkleProof, error)
+	// ConstructBatchedProof constructs a batched Merkle proof of the nodes from and including leftLvl, leftIdx, to and including rightLvl, rightIdx.
+	// That is, if leftLvl, or rightLvl, is not the leaf-level, then the proof is of the entire subtree from leftLvl at leftIdx to rightLvl at rightIdx
+	// Level 0 is the root and index 0 is the left-most node in a level.
+	ConstructBatchedProof(leftLvl int, leftIdx int, rightLvl int, rightIdx int) (BatchedMerkleProof, error)
+	// ValidateFromLeafs checks that the Merkle tree is correctly constructed based on all the leaf data
+	ValidateFromLeafs(leafData [][]byte) bool
+	// Validate checks that the Merkle tree is correctly constructed, based on the internal nodes
+	Validate() bool
+}
+
+type data struct {
+	// nodes start from root and go down left-to-right
+	// thus len(nodes[0]) = 1, len(nodes[1]) = 2, etc.
+	nodes [][]Node
+}
+
+type Node struct {
+	data [digestBytes]byte
+}
+
+// NewBareTree allocates that memory needed to construct a tree with a specific amount of leafs
+func NewBareTree(leafs int) MerkleTree {
```

Comment:
It is preferred to return the concrete struct type here, even if containing all private fields. This would avoid some type casts further in the code.

Reply:
Handled in commit [24bab9c](https://github.com/filecoin-project/go-data-segment/pull/2/commits/24bab9c51058322eca5afe1279c104d1a6ae1616)